File: 2021/day-07/go/main.go
package main

import (
	"fmt"
	"math"
	"sort"
)

func part1(data []float64) int {
	sort.Float64s(data)
	middle := float64(len(data)) / 2
	median := 0.0
	if len(data) % 2 == 1 {
		median = data[int(middle)]
	} else {
		median = math.Round((data[int(middle) - 1] + data[int(middle)]) / 2)
	}

	cost := 0
	for _, pos := range data {
		cost += int(math.Abs(pos - median))
	}
	return cost
}

// func part2(data) int {

// }

func main() {
	fmt.Println("Part One")
	fmt.Printf("  sample: %d\n", part1(sample))
	fmt.Printf("  input : %d\n", part1(input))
	// fmt.Println("Part Two")
	// fmt.Printf("  sample: %d\n", part2(sample))
	// fmt.Printf("  input : %d\n", part2(input))
}

File: 2021/day-07/rust/src/main.rs
mod input;

// sort data first please!!!
fn minimal_cost(data: &Vec<i64>, cost_fn: fn(&Vec<i64>, i64) -> i64) -> i64 {
	let left = data[0];
	let right = data[data.len() - 1];
	let mut cost = cost_fn(data, left);
	for p in (left + 1)..right {
		let c = cost_fn(data, p);
		if c > cost {
			break;
		};
		cost = c;
	}
	cost
}

fn cost1(data: &Vec<i64>, pos: i64) -> i64 {
	data.iter().map(|x| (x - pos).abs()).sum()
}

fn cost2(data: &Vec<i64>, pos: i64) -> i64 {
	data.iter()
		.map(|x| {
			let d = (x - pos).abs();
			d * (d + 1) / 2
		})
		.sum()
}

fn part1(data: &Vec<i64>) -> i64 {
	minimal_cost(data, cost1)
}

fn part2(data: &Vec<i64>) -> i64 {
	minimal_cost(data, cost2)
}

fn main() {
	// let mut sample: Vec<i64> = include_str!("sample")
	// 	.split(',')
	// 	.map(|n| n.parse().unwrap())
	// 	.collect();
	let mut sample = Vec::from(input::SAMPLE);
	sample.sort();
	let mut input = Vec::from(input::INPUT);
	input.sort();

	println!("=== Part 1 ===");
	println!("sample : {}", part1(&sample));
	println!("input  : {}", part1(&input));
	println!("=== Part 2 ===");
	println!("sample : {}", part2(&sample));
	println!("input  : {}", part2(&input));

	// === Part 1 ===
	// sample : 37
	// input  : 340987
	// === Part 2 ===
	// sample : 168
	// input  : 96987874
}
